perspective.el

```diff
@@ -950,6 +962,236 @@ perspective beginning with the given letter."
+    (unless target-file
+      (error "No target file specified"))
+    ;; overwrite the target file if:
+    ;; - the file does not exist, or
+    ;; - the file is not the one set in persp-state-default-file, or
+    ;; - the user called this function with a prefix argument, or
+    ;; - the user approves overwriting the file when prompted
+    (when (and (file-exists-p target-file)
+               (not (string-equal (if (and persp-state-default-file
+                                           (not (string-equal "" persp-state-default-file)))
+                                      (expand-file-name persp-state-default-file)
+                                    "")
+                                  target-file))
+               (not (or current-prefix-arg
+                        (yes-or-no-p "Target file exists. Overwrite? "))))
+      (error "persp-state-save cancelled"))
+    ;; if ivy-posframe is in use, get rid of its frame first
+    (when (and (fboundp 'posframe-delete) (boundp 'ivy-posframe-buffer))
+      (posframe-delete ivy-posframe-buffer))
+    ;; actually save
+    (persp-save)
+    (lexical-let ((state-complete (make-persp--state-complete
+                                   :files (persp--state-file-data)
+                                   :frames (persp--state-frame-data))))
+      (when (file-exists-p target-file)
+        (delete-file target-file))
+      (append-to-file (prin1-to-string state-complete) nil target-file))))
+
+;;;###autoload
+(defun persp-state-load (file)
+  "Restore the perspective state saved in FILE.
+
+FILE defaults to the value of persp-state-default-file if it is
+set.
+
+Frames are restored, along with each frame's perspective list.
+Each perspective's buffer list and window layout are also
+restored."
+  (interactive (list
+                (read-file-name "Restore perspective state from file: "
+                                persp-state-default-file
+                                persp-state-default-file)))
+  (unless (file-exists-p file)
+    (error "File not found"))
+  (persp-mode 1)
+  (lexical-let ((tmp-persp-name (format "%04x%04x" (random (expt 16 4)) (random (expt 16 4))))
+                (frame-count 0)
+                (state-complete (read-from-whole-string
+                                 (with-temp-buffer
+                                   (insert-file-contents file)
+                                   (buffer-string)))))
+    ;; open all files in a temporary perspective to avoid polluting "main"
+    (persp-switch tmp-persp-name)
+    (cl-loop for file in (persp--state-complete-files state-complete) do
+             (when (file-exists-p file)
+               (find-file file)))
+    ;; iterate over the frames
+    (cl-loop for frame in (persp--state-complete-frames state-complete) do
+             (incf frame-count)
+             (when (> frame-count 1)
+               (make-frame-command))
+             ;; backwards compatibility with older persp-state files: the frame
```

Comment:
I agonized over this. The backwards compatibility was quite useful to me personally, and I've been suggesting that people use this unreleased patch, and several have kindly tested it. I'd rather not leave them with useless state files after this goes out. On the other hand, you're right: this adds ugly complexity.

In the end, I think this should go out with the backwards compatibility mode. It will help early adopters by not breaking their existing state files, and they will quickly generate new working state files as they use perspective. I'll remove the backwards compatibility layer in the (not distant) future, and left myself a comment as a reminder.